osquery/tables/system/windows/registry.cpp

```diff
@@ -266,33 +269,145 @@ void queryKey(const std::string& keyPath, QueryData& results) {
-  /// By default, we display all HIVEs
-  if ((context.constraints["key"].exists(EQUALS) &&
-       context.constraints["key"].getAll(EQUALS).size() > 0)) {
-    rKeys = context.constraints["key"].getAll(EQUALS);
-    shouldWarnLocalUsers = true;
-  } else {
-    for (auto& h : kRegistryHives) {
-      rKeys.insert(h.first);
+void populateDefaultKeys(std::set<std::string>& rKeys) {
+  for (const auto& hive : kRegistryHives) {
+    rKeys.insert(hive.first);
+  }
+}
+
+void replaceKeysWithSubkeys(std::set<std::string>& rKeys) {
+  std::set<std::string> newKeys{};
+  for (const auto& key : rKeys) {
+    QueryData regResults;
+    queryKey(key, regResults);
+    for (const auto& r : regResults) {
+      if (r.at("type") == "subkey") {
+        newKeys.insert(r.at("path"));
+      }
+    }
+  }
+  rKeys = newKeys;
+}
+
+void appendSubkeyToKeys(const std::string& subkey,
+                        std::set<std::string>& rKeys) {
+  std::set<std::string> newKeys{};
+  for (auto& key : rKeys) {
+    newKeys.insert(key + kRegSep + subkey);
+  }
+  rKeys = newKeys;
+}
+
+Status populateAllKeysRecursive(std::set<std::string>& rKeys,
+                                int currDepth,
+                                int maxDepth) {
+  std::set<std::string> subkeys{};
+
+  if (currDepth > maxDepth) {
+    return Status(
+        1, "Max recursive depth (" + std::to_string(maxDepth) + ") reached");
+  }
+
+  for (const auto& key : rKeys) {
+    QueryData regResults;
+    queryKey(key, regResults);
+    for (const auto& r : regResults) {
+      if (r.at("type") == "subkey") {
+        subkeys.insert(r.at("path"));
+      }
+    }
+  }
+
+  if (!(subkeys.size() == 0)) {
+    auto status = populateAllKeysRecursive(subkeys, ++currDepth);
+    if (!status.ok()) {
+      return status;
+    }
+    rKeys.insert(subkeys.begin(), subkeys.end());
+  }
+
+  return Status(0, "OK");
+}
+
+Status resolveRegistryGlobs(const std::string& pattern,
+                            std::set<std::string>& results) {
+  auto pathElems = osquery::split(pattern, kRegSep);
+
+  // Pattern is '%%', grab everything
+  // Note that if '%%' is present but not at the end of the pattern
+  // , then it is treated like a single glob
+  if (boost::ends_with(pathElems[0], kRegRecursiveGlob) && pathElems.size() == 1) {
+    populateDefaultKeys(results);
+    return populateAllKeysRecursive(results);
+  }
+  
+  // Special handling to insert default keys when glob present in first elem
+  if (pathElems[0].find(kRegSingleGlob) != std::string::npos) {
+    populateDefaultKeys(results);
+    pathElems.erase(pathElems.begin());
+  }
+  else {
+    results.insert(pathElems[0]);
+    pathElems.erase(pathElems.begin());
+  }
+
+  for (const auto& elem : pathElems) {
+     // We only care about  a recursive glob if it comes at the end of the pattern
+     // i.e. 'HKEY_LOCAL_MACHINE\SOFTWARE\%%'
+    if (boost::ends_with(elem, kRegRecursiveGlob) && &elem == &pathElems.back()) {
```

Comment:
yes, this seemed to be one of the possible ways to verify that the current elem is also the final element:
```
&elem == &pathElems.back()
```

Is there a recommended/better way to do this?